Sources/NIOCore/NIOOnLoopSendable.swift

```diff
@@ -0,0 +1,78 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+public struct NIOOnLoopSendable<Value>: @unchecked Sendable {
+    @usableFromInline
+    let _eventLoop: EventLoop
+
+    @usableFromInline
+    var _value: Value
+
+    @inlinable
+    public init(_ value: Value, eventLoop: EventLoop) {
+        eventLoop.preconditionInEventLoop()
+        self._eventLoop = eventLoop
+        self._value = value
+    }
+
+    @inlinable
+    public var value: Value {
+        get {
+            self._eventLoop.preconditionInEventLoop()
+            return self._value
+        }
+        set {
+            self._eventLoop.preconditionInEventLoop()
+            self._value = newValue
+        }
+    }
+}
+
+public class NIOOnLoopSendableBox<Value>: @unchecked Sendable {
+    @usableFromInline
+    let _eventLoop: EventLoop
+
+    @usableFromInline
+    var _value: Value
+
+    @inlinable
+    public init(_ value: Value, eventLoop: EventLoop) {
+        eventLoop.preconditionInEventLoop()
+        self._eventLoop = eventLoop
+        self._value = value
+    }
+
+    public init<NonOptionalValue>(
+        valueType: NonOptionalValue.Type = NonOptionalValue.self,
+        eventLoop: EventLoop
+    ) where Optional<NonOptionalValue> == Value {
+        // We are deliberately _not_ checking if we're on `EventLoop` here because we're only putting in the `.none`
+        // value of `Optional`.
+        self._eventLoop = eventLoop
+        self._value = .none
+    }
```

Comment:
So this can be used to initialise a variable from one thread but read/write it from another thread without any memory barrier. This sound as if the compiler might be allowed to reorder this. cc: @Lukasa 

In addition, we need to make it clear that we initialise `value` to nil. Preferable in the symbol name or at least in the docs.

Reply:
There's no reordering risk here: you cannot publish a reference to `self` until it is fully initialized, so we cannot publish `self` until such time as this object is properly constructed.

Reply:
@dnadoba This isn't quite correct. The value is fully initialised before `init` returns. This reference isn't and cannot be published before `init` returns and is guaranteed to be local to just one thread up until after `init` has returned. And crucially, it is never read or written anymore by anybody except for the EL thread.

To then ever read or write it, you need to get it onto _the_ EventLoop (unless you're already there) which requires a memory barrier because EventLoops have to guarantee total ordering.

Reply:
Thanks for explaining! I think it would be great to explain this also in the body of the function.

Reply:
I'll expand the comment.